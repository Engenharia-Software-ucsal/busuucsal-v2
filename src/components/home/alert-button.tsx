import {
  Select,
  SelectBackdrop,
  SelectContent,
  SelectDragIndicator,
  SelectDragIndicatorWrapper,
  SelectItem,
  SelectPortal,
  SelectTrigger,
} from "@/components/ui/select";

import { Text } from "@/components/ui/text";
import {
  Toast,
  ToastDescription,
  ToastTitle,
  useToast,
} from "@/components/ui/toast";

export enum AlertType {
  BUS_DELAY = "bus_delay",
  BUS_BROKEN = "bus_broken",
  BUS_FULL = "bus_full",
}

export function AlertButton() {
  const toast = useToast();

  function handleSelectAlert(value: AlertType) {
    toast.show({
      id: "alert",
      placement: "top",
      containerStyle: {
        marginTop: 50,
      },
      render({ id }) {
        return (
          <Toast nativeID={id} variant="outline" action="info">
            <ToastTitle>Alerta emitido !</ToastTitle>
            <ToastDescription>
              Avisaremos outros usu√°rios sobre o seu alerta
            </ToastDescription>
          </Toast>
        );
      },
    });
  }

  return (
    <Select
      onValueChange={(value) => {
        handleSelectAlert(value as AlertType);
      }}
    >
      <SelectTrigger variant="outline" size="lg">
        <Text className="p-2 text-lg">Notificar alerta</Text>
      </SelectTrigger>
      <SelectPortal>
        <SelectBackdrop />
        <SelectContent>
          <SelectDragIndicatorWrapper>
            <SelectDragIndicator />
          </SelectDragIndicatorWrapper>
          <SelectItem label="Onibus atrasou" value={AlertType.BUS_DELAY} />
          <SelectItem label="Onbius quebrou" value={AlertType.BUS_BROKEN} />
          <SelectItem label="Onibus cheio" value={AlertType.BUS_FULL} />
        </SelectContent>
      </SelectPortal>
    </Select>
  );
}


export default AlertButton;