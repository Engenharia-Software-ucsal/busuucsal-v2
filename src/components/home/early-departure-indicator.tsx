import React from "react";

import { Heading } from "@/components/ui/heading";
import { Box } from "@/components/ui/box";
import { VStack } from "@/components/ui/vstack";
import { Text } from "@/components/ui/text";

import { useAtomValue } from "jotai";
import {
  distanceToNextDepartureAtom,
  earlyDepartureTimeAtom,
} from "@/atoms/itinenary";
import { formattedDateAtom } from "@/atoms/date";
import { HStack } from "@/components/ui/hstack";
import { Icon } from "@/components/ui/icon";
import { BusFront, Calendar } from "lucide-react-native";

export function EarlyDepartureIndicator() {
  const todayTitle = useAtomValue(formattedDateAtom);
  const earlyNextDeparture = useAtomValue(earlyDepartureTimeAtom);
  const distanceToNextDeparture = useAtomValue(distanceToNextDepartureAtom);

  return (
    <>
      <HStack className="items-center gap-2">
        <Icon as={Calendar} size="md" />
        <Heading size="xl" className="flex justify-center items-center">
          {todayTitle}
        </Heading>
      </HStack>

      <Box className="mt-10 w-[250px] h-[250px] border border-blue-400 rounded-full">
        <VStack space="lg" className="justify-center items-center flex-1">
          {earlyNextDeparture && <Text size="xl">Próxima saída</Text>}

          <Text size="3xl">{earlyNextDeparture ?? "Sem horários"}</Text>

          <Icon as={BusFront} size={30 as any} />

          {distanceToNextDeparture && (
            <Text>
              <Text className="font-bold">{distanceToNextDeparture}</Text>
            </Text>
          )}
        </VStack>
      </Box>
    </>
  );
}
