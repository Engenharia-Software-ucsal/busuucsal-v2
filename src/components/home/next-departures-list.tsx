import { nextDeparturesAtom } from "@/atoms/itinenary";
import { Box } from "@/components/ui/box";
import { Card } from "@/components/ui/card";
import { Heading } from "@/components/ui/heading";
import { HStack } from "@/components/ui/hstack";
import { Icon } from "@/components/ui/icon";
import { VStack } from "@/components/ui/vstack";
import { useAtomValue } from "jotai";
import { map, tail } from "lodash";
import { Clock } from "lucide-react-native";
import React from "react";
import { FlatList } from "react-native";
import { Text } from "../ui/text";

export function NextDepartureListIem({
  departure,
  arrival,
}: {
  departure: string;
  arrival: string;
}) {
  return (
    <Box className="flex-row justify-between">
      <Text>{departure}</Text>
      <Text>{arrival}</Text>
    </Box>
  );
}

export function NextDeparturesList() {
  const nextDepartures = useAtomValue(nextDeparturesAtom);

  const parsedNextDeparturesList = tail(
    map(nextDepartures, (item) => ({
      departure: item.departure.concat(" h"),
      arrival: item.arrival.concat(" h"),
    })),
  );

  return (
    <>
      <HStack className="mt-20 items-center" space="sm">
        <Icon as={Clock} size="xl" />
        <Heading
          testID="next_departures_title"
          size="xl"
          className="flex justify-center items-center  "
        >
          Próximos horários
        </Heading>
      </HStack>
      <Box className="w-full px-10 mt-10 mb-14">
        <Card size="lg" className="w-full max-h-[200px]">
          <VStack space="2xl">
            <Box className="flex-row justify-between">
              <Heading>Saída</Heading>
              <Heading>Chegada</Heading>
            </Box>

            <FlatList
              testID="next_departures_list"
              style={{ maxHeight: 100 }}
              keyExtractor={({ departure, arrival }, index) =>
                `${index}-${departure}-${arrival}`
              }
              data={parsedNextDeparturesList}
              ItemSeparatorComponent={() => <Box className="mt-4" />}
              renderItem={({ item }) => (
                <NextDepartureListIem
                  arrival={item.arrival}
                  departure={item.departure}
                />
              )}
              ListEmptyComponent={() => (
                <Box className="flex justify-center items-center w-full h-[100px]">
                  <Text>Nenhum horário disponível</Text>
                </Box>
              )}
            />
          </VStack>
        </Card>
      </Box>
    </>
  );
}
