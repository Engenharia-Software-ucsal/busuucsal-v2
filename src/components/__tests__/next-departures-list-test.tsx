import * as React from "react";
import renderer from "react-test-renderer";
import { NextDeparturesList } from "../home/next-departures-list";

describe("<NextDeparturesList />", () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it("renders NextDeparturesList component", () => {
    const tree = renderer.create(<NextDeparturesList />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render the list of next departures", () => {
    const tree = renderer.create(<NextDeparturesList />);
  });
});
