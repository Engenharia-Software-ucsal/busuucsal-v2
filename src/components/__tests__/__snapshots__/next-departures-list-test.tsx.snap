// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NextDeparturesList /> renders NextDeparturesList component 1`] = `
[
  <View
    className="flex-row gap-2 mt-20 items-center"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="24"
      bbWidth="24"
      focusable={false}
      height={24}
      meetOrSlice={0}
      minX={0}
      minY={0}
      role="img"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      vbHeight={24}
      vbWidth={24}
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        propList={
          [
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth="2"
      >
        <RNSVGCircle
          cx="12"
          cy="12"
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          r="10"
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="2"
        />
        <RNSVGPath
          d="M12 6 12 12 16 14"
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "type": 2,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth="2"
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      accessibilityRole="header"
      style={
        {
          "fontSize": 37.44,
          "fontWeight": "bold",
          "marginVertical": 32,
        }
      }
      testID="next_departures_title"
    >
      Próximos horários
    </Text>
  </View>,
  <View
    className="w-full px-10 mt-10 mb-14"
  >
    <View
      className="p-6 rounded-xl bg-background-0 w-full max-h-[200px]"
    >
      <View
        className="flex-col gap-6"
      >
        <View
          className="flex-row justify-between"
        >
          <Text
            accessibilityRole="header"
            style={
              {
                "fontSize": 32,
                "fontWeight": "bold",
                "marginVertical": 42.56,
              }
            }
          >
            Saída
          </Text>
          <Text
            accessibilityRole="header"
            style={
              {
                "fontSize": 32,
                "fontWeight": "bold",
                "marginVertical": 42.56,
              }
            }
          >
            Chegada
          </Text>
        </View>
        <RCTScrollView
          ItemSeparatorComponent={[Function]}
          ListEmptyComponent={[Function]}
          data={
            [
              {
                "arrival": "10:20 h",
                "departure": "9:55 h",
              },
              {
                "arrival": "11:10 h",
                "departure": "10:45 h",
              },
              {
                "arrival": "12:00 h",
                "departure": "11:35 h",
              },
              {
                "arrival": "12:50 h",
                "departure": "12:25 h",
              },
              {
                "arrival": "18:10 h",
                "departure": "17:10 h",
              },
              {
                "arrival": "18:50 h",
                "departure": "18:30 h",
              },
              {
                "arrival": "20:35 h",
                "departure": "20:10 h",
              },
              {
                "arrival": "21:50 h",
                "departure": "21:25 h",
              },
              {
                "arrival": "22:10 h",
                "departure": "21:50 h",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={0.0001}
          stickyHeaderIndices={[]}
          style={
            {
              "maxHeight": 100,
            }
          }
          viewabilityConfigCallbackPairs={[]}
        >
          <View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  9:55 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  10:20 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  10:45 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  11:10 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  11:35 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  12:00 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  12:25 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  12:50 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  17:10 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  18:10 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  18:30 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  18:50 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  20:10 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  20:35 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  21:25 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  21:50 h
                </Text>
              </View>
              <View
                className="mt-4"
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                className="flex-row justify-between"
              >
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  21:50 h
                </Text>
                <Text
                  className="text-typography-700 font-body text-base"
                >
                  22:10 h
                </Text>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>,
]
`;
